core/src/main/java/org/neo4j/graphalgo/core/utils/DisjointSetStruct.java

```diff
@@ -0,0 +1,72 @@
+package org.neo4j.graphalgo.core.utils;
+
+/**
+ * A structure to manage disjoint sets of ints e.g. nodeIds
+ *
+ * @author mknblch
+ */
+public class DisjointSetStruct {
+
+    private final int[] parent;
+    private final int[] depth;
+
+    /**
+     *
+     * @param capacity the capacity (maximum node id)
+     */
+    public DisjointSetStruct(int capacity) {
+        parent = new int[capacity];
```

Comment:
The parent array is more like an index-to-parentIndex map. We can switch to an implementation based on linked lists which could grow on-demand but as usual decrease performance. We might consider having both implementations and some kind of configuration to choose one.